File: backend/app/api/v1/accounting.py
"""
Accounting API Module

This module provides REST API endpoints for accounting/session statistics
and reporting functionality.
"""

from typing import Optional, List
from datetime import datetime
from fastapi import APIRouter, Depends, HTTPException, Query, status
from fastapi.responses import JSONResponse

from app.core.auth import get_current_user
from app.core.database import get_db
from app.core.exceptions import NotFoundError, ValidationError, BusinessLogicError
from app.models import User
from app.repositories.accounting import (
    AccountingRepository,
    UserTrafficSummaryRepository,
    NasTrafficSummaryRepository
)
from app.services.accounting import (
    AccountingService,
    UserTrafficSummaryService,
    NasTrafficSummaryService
)
from app.schemas.accounting import (
    RadAcctResponse, AccountingQuery, AccountingQueryFilters,
    PaginatedAccountingResponse, AccountingOverview,
    HourlyTrafficReport,
    NasUsageReport, CustomQueryResult, MaintenanceResult,
    PaginatedTopUsersResponse, UserTrafficSummaryResponse,
    NasTrafficSummaryResponse, AccountingTimeRangeEnum,
    CustomQueryRequest, MaintenanceRequest
)
from sqlalchemy.orm import Session

router = APIRouter()


def get_accounting_service(db: Session = Depends(get_db)) -> AccountingService:
    """Get accounting service instance"""
    repository = AccountingRepository(db)
    return AccountingService(repository)


def get_user_traffic_summary_service(db: Session = Depends(get_db)) -> UserTrafficSummaryService:
    """Get user traffic summary service instance"""
    repository = UserTrafficSummaryRepository(db)
    return UserTrafficSummaryService(repository)


def get_nas_traffic_summary_service(db: Session = Depends(get_db)) -> NasTrafficSummaryService:
    """Get NAS traffic summary service instance"""
    repository = NasTrafficSummaryRepository(db)
    return NasTrafficSummaryService(repository)


# =====================================================================
# Session Management Endpoints
# =====================================================================

@router.get("/sessions", response_model=PaginatedAccountingResponse)
async def get_accounting_sessions(
    page: int = Query(1, ge=1, description="Page number"),
    page_size: int = Query(20, ge=1, le=100, description="Items per page"),
    sort_field: Optional[str] = Query(
        "acctstarttime", description="Sort field"),
    sort_order: Optional[str] = Query(
        "desc", regex="^(asc|desc)$", description="Sort order"),
    # Filters
    username: Optional[str] = Query(None, description="Filter by username"),
    groupname: Optional[str] = Query(None, description="Filter by group name"),
    nasipaddress: Optional[str] = Query(
        None, description="Filter by NAS IP address"),
    framedipaddress: Optional[str] = Query(
        None, description="Filter by framed IP address"),
    callingstationid: Optional[str] = Query(
        None, description="Filter by calling station ID"),
    servicetype: Optional[str] = Query(
        None, description="Filter by service type"),
    time_range: Optional[AccountingTimeRangeEnum] = Query(
        None, description="Predefined time range"),
    start_date: Optional[datetime] = Query(
        None, description="Start date filter"),
    end_date: Optional[datetime] = Query(None, description="End date filter"),
    active_only: Optional[bool] = Query(
        None, description="Show only active sessions"),
    # Current user dependency
    current_user: User = Depends(get_current_user),
    service: AccountingService = Depends(get_accounting_service)
):
    """Get paginated accounting sessions with filtering"""
    try:
        # Build query filters
        filters = AccountingQueryFilters(
            username=username,
            groupname=groupname,
            nasipaddress=nasipaddress,
            framedipaddress=framedipaddress,
            callingstationid=callingstationid,
            servicetype=servicetype,
            time_range=time_range,
            start_date=start_date,
            end_date=end_date,
            active_only=active_only
        )

        # Create query
        query = AccountingQuery(
            page=page,
            page_size=page_size,
            sort_field=sort_field,
            sort_order=sort_order,
            filters=filters
        )

        # Get sessions
        return await service.get_accounting_sessions(query)

    except ValidationError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
    except BusinessLogicError as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.get("/sessions/{session_id}", response_model=RadAcctResponse)
async def get_session_by_id(
    session_id: int,
    current_user: User = Depends(get_current_user),
    service: AccountingService = Depends(get_accounting_service)
):
    """Get accounting session by ID"""
    try:
        return await service.get_session_by_id(session_id)
    except NotFoundError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except BusinessLogicError as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.get("/sessions/active", response_model=PaginatedAccountingResponse)
async def get_active_sessions(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    nas_ip: Optional[str] = Query(None, description="Filter by NAS IP"),
    username: Optional[str] = Query(None, description="Filter by username"),
    current_user: User = Depends(get_current_user),
    service: AccountingService = Depends(get_accounting_service)
):
    """Get active sessions"""
    try:
        return await service.get_active_sessions(
            page=page,
            page_size=page_size,
            nas_ip=nas_ip,
            username=username
        )
    except ValidationError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
    except BusinessLogicError as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.get("/sessions/user/{username}", response_model=PaginatedAccountingResponse)
async def get_user_sessions(
    username: str,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    date_from: Optional[datetime] = Query(None),
    date_to: Optional[datetime] = Query(None),
    current_user: User = Depends(get_current_user),
    service: AccountingService = Depends(get_accounting_service)
):
    """Get sessions for a specific user"""
    try:
        return await service.get_user_sessions(
            username=username,
            page=page,
            page_size=page_size,
            date_from=date_from,
            date_to=date_to
        )
    except ValidationError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
    except BusinessLogicError as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


# =====================================================================
# Statistics and Overview Endpoints
# =====================================================================

@router.get("/overview", response_model=AccountingOverview)
async def get_accounting_overview(
    time_range: Optional[AccountingTimeRangeEnum] = Query(None),
    date_from: Optional[datetime] = Query(None),
    date_to: Optional[datetime] = Query(None),
    # Filters
    username: Optional[str] = Query(None),
    groupname: Optional[str] = Query(None),
    nasipaddress: Optional[str] = Query(None),
    current_user: User = Depends(get_current_user),
    service: AccountingService = Depends(get_accounting_service)
):
    """Get comprehensive accounting overview"""
    try:
        filters = None
        if username or groupname or nasipaddress:
            filters = AccountingQueryFilters(
                username=username,
                groupname=groupname,
                nasipaddress=nasipaddress
            )

        return await service.get_accounting_overview(
            time_range=time_range,
            date_from=date_from,
            date_to=date_to,
            filters=filters
        )
    except ValidationError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
    except BusinessLogicError as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


# =====================================================================
# Reports Endpoints
# =====================================================================

@router.get("/reports/top-users", response_model=PaginatedTopUsersResponse)
async def get_top_users_report(
    limit: int = Query(10, ge=1, le=100, description="Number of top users"),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    date_from: Optional[datetime] = Query(None),
    date_to: Optional[datetime] = Query(None),
    current_user: User = Depends(get_current_user),
    service: AccountingService = Depends(get_accounting_service)
):
    """Get top users by traffic consumption"""
    try:
        return await service.get_top_users_report(
            limit=limit,
            page=page,
            page_size=page_size,
            date_from=date_from,
            date_to=date_to
        )
    except BusinessLogicError as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.get("/reports/hourly-traffic", response_model=List[HourlyTrafficReport])
async def get_hourly_traffic_report(
    date_from: Optional[datetime] = Query(None),
    date_to: Optional[datetime] = Query(None),
    current_user: User = Depends(get_current_user),
    service: AccountingService = Depends(get_accounting_service)
):
    """Get hourly traffic distribution report"""
    try:
        return await service.get_hourly_traffic_report(
            date_from=date_from,
            date_to=date_to
        )
    except BusinessLogicError as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.get("/reports/nas-usage", response_model=List[NasUsageReport])
async def get_nas_usage_report(
    date_from: Optional[datetime] = Query(None),
    date_to: Optional[datetime] = Query(None),
    current_user: User = Depends(get_current_user),
    service: AccountingService = Depends(get_accounting_service)
):
    """Get NAS usage statistics report"""
    try:
        return await service.get_nas_usage_report(
            date_from=date_from,
            date_to=date_to
        )
    except BusinessLogicError as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


# =====================================================================
# Custom Queries and Maintenance
# =====================================================================

@router.post("/custom-query", response_model=CustomQueryResult)
async def execute_custom_query(
    request: CustomQueryRequest,
    current_user: User = Depends(get_current_user),
    service: AccountingService = Depends(get_accounting_service)
):
    """Execute custom accounting query"""
    try:
        return await service.execute_custom_query(
            query_sql=request.query_sql,
            parameters=request.parameters,
            limit=request.limit
        )
    except ValidationError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
    except BusinessLogicError as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.post("/maintenance/cleanup", response_model=MaintenanceResult)
async def cleanup_old_sessions(
    request: MaintenanceRequest,
    current_user: User = Depends(get_current_user),
    service: AccountingService = Depends(get_accounting_service)
):
    """Clean up old accounting sessions"""
    try:
        return await service.cleanup_old_sessions(
            days_old=request.days_old,
            dry_run=request.dry_run
        )
    except ValidationError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
    except BusinessLogicError as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


# =====================================================================
# Traffic Summary Endpoints
# =====================================================================

@router.get("/traffic-summary/user/{username}", response_model=List[UserTrafficSummaryResponse])
async def get_user_traffic_summary(
    username: str,
    date_from: Optional[datetime] = Query(None),
    date_to: Optional[datetime] = Query(None),
    current_user: User = Depends(get_current_user),
    service: UserTrafficSummaryService = Depends(
        get_user_traffic_summary_service)
):
    """Get traffic summary for a user"""
    try:
        return await service.get_user_traffic_summary(
            username=username,
            date_from=date_from.date() if date_from else None,
            date_to=date_to.date() if date_to else None
        )
    except BusinessLogicError as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.get("/traffic-summary/nas/{nasipaddress}", response_model=List[NasTrafficSummaryResponse])
async def get_nas_traffic_summary(
    nasipaddress: str,
    date_from: Optional[datetime] = Query(None),
    date_to: Optional[datetime] = Query(None),
    current_user: User = Depends(get_current_user),
    service: NasTrafficSummaryService = Depends(
        get_nas_traffic_summary_service)
):
    """Get traffic summary for a NAS"""
    try:
        return await service.get_nas_traffic_summary(
            nasipaddress=nasipaddress,
            date_from=date_from.date() if date_from else None,
            date_to=date_to.date() if date_to else None
        )
    except BusinessLogicError as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


# =====================================================================
# Health Check
# =====================================================================

@router.get("/health")
async def health_check():
    """Health check endpoint for accounting module"""
    return JSONResponse(
        content={
            "status": "healthy",
            "module": "accounting",
            "timestamp": datetime.utcnow().isoformat()
        }
    )


# Export router
__all__ = ["router"]

File: frontend/src/services/accounting.ts
/**
 * Accounting API Service
 * 
 * API service for accounting/session statistics operations
 */

import { apiClient } from '@/services/api'
import type {
  AccountingSession,
  AccountingQuery,
  AccountingOverview,
  TopUsersReport,
  HourlyTrafficReport,
  NasUsageReport,
  UserTrafficSummary,
  NasTrafficSummary,
  PaginatedAccountingResponse,
  PaginatedTopUsersResponse,
  CustomQueryRequest,
  CustomQueryResult,
  MaintenanceRequest,
  MaintenanceResult,
  AccountingApiParams,
  ReportFilters,
  AccountingTimeRangeEnum
} from '@/types/accounting'

/**
 * Accounting API Service Class
 */
export class AccountingService {
  private readonly baseUrl = '/accounting'

  // =====================================================================
  // Session Management
  // =====================================================================

  /**
   * Get paginated accounting sessions with filtering
   */
  async getSessions(query: AccountingQuery): Promise<PaginatedAccountingResponse> {
    const params = new URLSearchParams()
    
    // Pagination
    params.append('page', query.page.toString())
    params.append('page_size', query.page_size.toString())
    
    // Sorting
    if (query.sort_field) params.append('sort_field', query.sort_field)
    if (query.sort_order) params.append('sort_order', query.sort_order)
    
    // Filters
    if (query.filters) {
      Object.entries(query.filters).forEach(([key, value]) => {
        if (value !== null && value !== undefined && value !== '') {
          params.append(key, value.toString())
        }
      })
    }

    const response = await apiClient.get(`${this.baseUrl}/sessions?${params.toString()}`)
    return response.data
  }

  /**
   * Get accounting session by ID
   */
  async getSessionById(sessionId: number): Promise<AccountingSession> {
    const response = await apiClient.get(`${this.baseUrl}/sessions/${sessionId}`)
    return response.data
  }

  /**
   * Get active sessions
   */
  async getActiveSessions(params: AccountingApiParams['activeSessions'] = {}): Promise<PaginatedAccountingResponse> {
    const searchParams = new URLSearchParams()
    
    if (params.page) searchParams.append('page', params.page.toString())
    if (params.page_size) searchParams.append('page_size', params.page_size.toString())
    if (params.nas_ip) searchParams.append('nas_ip', params.nas_ip)
    if (params.username) searchParams.append('username', params.username)

    const response = await apiClient.get(`${this.baseUrl}/sessions/active?${searchParams.toString()}`)
    return response.data
  }

  /**
   * Get sessions for a specific user
   */
  async getUserSessions(
    username: string,
    params: { page?: number; page_size?: number; date_from?: string; date_to?: string } = {}
  ): Promise<PaginatedAccountingResponse> {
    const searchParams = new URLSearchParams()
    
    if (params.page) searchParams.append('page', params.page.toString())
    if (params.page_size) searchParams.append('page_size', params.page_size.toString())
    if (params.date_from) searchParams.append('date_from', params.date_from)
    if (params.date_to) searchParams.append('date_to', params.date_to)

    const response = await apiClient.get(`${this.baseUrl}/sessions/user/${username}?${searchParams.toString()}`)
    return response.data
  }

  // =====================================================================
  // Statistics and Overview
  // =====================================================================

  /**
   * Get comprehensive accounting overview
   */
  async getOverview(params: AccountingApiParams['overview'] = {}): Promise<AccountingOverview> {
    const searchParams = new URLSearchParams()
    
    if (params.time_range) searchParams.append('time_range', params.time_range)
    if (params.date_from) searchParams.append('date_from', params.date_from)
    if (params.date_to) searchParams.append('date_to', params.date_to)
    if (params.filters) {
      Object.entries(params.filters).forEach(([key, value]) => {
        if (value !== null && value !== undefined && value !== '') {
          searchParams.append(key, value.toString())
        }
      })
    }

    const response = await apiClient.get(`${this.baseUrl}/overview?${searchParams.toString()}`)
    return response.data
  }

  // =====================================================================
  // Reports
  // =====================================================================

  /**
   * Get top users by traffic consumption
   */
  async getTopUsersReport(params: AccountingApiParams['topUsers'] = {}): Promise<PaginatedTopUsersResponse> {
    const searchParams = new URLSearchParams()
    
    if (params.limit) searchParams.append('limit', params.limit.toString())
    if (params.page) searchParams.append('page', params.page.toString())
    if (params.page_size) searchParams.append('page_size', params.page_size.toString())
    if (params.date_from) searchParams.append('date_from', params.date_from)
    if (params.date_to) searchParams.append('date_to', params.date_to)

    const response = await apiClient.get(`${this.baseUrl}/reports/top-users?${searchParams.toString()}`)
    return response.data
  }

  /**
   * Get hourly traffic distribution report
   */
  async getHourlyTrafficReport(filters: ReportFilters = {}): Promise<HourlyTrafficReport[]> {
    const searchParams = new URLSearchParams()
    
    if (filters.date_from) searchParams.append('date_from', filters.date_from)
    if (filters.date_to) searchParams.append('date_to', filters.date_to)

    const response = await apiClient.get(`${this.baseUrl}/reports/hourly-traffic?${searchParams.toString()}`)
    return response.data
  }

  /**
   * Get NAS usage statistics report
   */
  async getNasUsageReport(filters: ReportFilters = {}): Promise<NasUsageReport[]> {
    const searchParams = new URLSearchParams()
    
    if (filters.date_from) searchParams.append('date_from', filters.date_from)
    if (filters.date_to) searchParams.append('date_to', filters.date_to)

    const response = await apiClient.get(`${this.baseUrl}/reports/nas-usage?${searchParams.toString()}`)
    return response.data
  }

  // =====================================================================
  // Custom Queries and Maintenance
  // =====================================================================

  /**
   * Execute custom accounting query
   */
  async executeCustomQuery(request: CustomQueryRequest): Promise<CustomQueryResult> {
    const response = await apiClient.post(`${this.baseUrl}/custom-query`, request)
    return response.data
  }

  /**
   * Clean up old accounting sessions
   */
  async cleanupOldSessions(request: MaintenanceRequest): Promise<MaintenanceResult> {
    const response = await apiClient.post(`${this.baseUrl}/maintenance/cleanup`, request)
    return response.data
  }

  // =====================================================================
  // Traffic Summaries
  // =====================================================================

  /**
   * Get traffic summary for a user
   */
  async getUserTrafficSummary(
    username: string,
    dateFrom?: string,
    dateTo?: string
  ): Promise<UserTrafficSummary[]> {
    const searchParams = new URLSearchParams()
    
    if (dateFrom) searchParams.append('date_from', dateFrom)
    if (dateTo) searchParams.append('date_to', dateTo)

    const response = await apiClient.get(
      `${this.baseUrl}/traffic-summary/user/${username}?${searchParams.toString()}`
    )
    return response.data
  }

  /**
   * Get traffic summary for a NAS
   */
  async getNasTrafficSummary(
    nasIpAddress: string,
    dateFrom?: string,
    dateTo?: string
  ): Promise<NasTrafficSummary[]> {
    const searchParams = new URLSearchParams()
    
    if (dateFrom) searchParams.append('date_from', dateFrom)
    if (dateTo) searchParams.append('date_to', dateTo)

    const response = await apiClient.get(
      `${this.baseUrl}/traffic-summary/nas/${nasIpAddress}?${searchParams.toString()}`
    )
    return response.data
  }

  // =====================================================================
  // Session Management Operations
  // =====================================================================

  /**
   * Disconnect a session (for active sessions)
   */
  async disconnectSession(sessionId: number): Promise<void> {
    await apiClient.post(`${this.baseUrl}/sessions/${sessionId}/disconnect`)
  }

  /**
   * Terminate a session (force stop)
   */
  async terminateSession(sessionId: number): Promise<void> {
    await apiClient.post(`${this.baseUrl}/sessions/${sessionId}/terminate`)
  }

  // =====================================================================
  // Export Functions
  // =====================================================================

  /**
   * Export sessions data
   */
  async exportSessions(params: { filters?: any; format?: string }): Promise<Blob> {
    const searchParams = new URLSearchParams()
    
    if (params.format) searchParams.append('format', params.format)
    if (params.filters) {
      Object.entries(params.filters).forEach(([key, value]) => {
        if (value !== null && value !== undefined && value !== '') {
          searchParams.append(key, value.toString())
        }
      })
    }

    const response = await apiClient.get(
      `${this.baseUrl}/sessions/export?${searchParams.toString()}`,
      { responseType: 'blob' }
    )
    
    // Create download
    const blob = new Blob([response.data], { 
      type: response.headers['content-type'] || 'application/octet-stream' 
    })
    
    const url = window.URL.createObjectURL(blob)
    const link = document.createElement('a')
    link.href = url
    link.download = `accounting-sessions-${new Date().toISOString().split('T')[0]}.${params.format || 'csv'}`
    document.body.appendChild(link)
    link.click()
    document.body.removeChild(link)
    window.URL.revokeObjectURL(url)

    return blob
  }

  /**
   * Export report data
   */
  async exportReport(params: AccountingApiParams['export']): Promise<Blob> {
    const searchParams = new URLSearchParams()
    
    searchParams.append('type', params.type)
    searchParams.append('format', params.format)
    
    if (params.filters) {
      Object.entries(params.filters).forEach(([key, value]) => {
        if (value !== null && value !== undefined && value !== '') {
          searchParams.append(key, value.toString())
        }
      })
    }

    const response = await apiClient.get(
      `${this.baseUrl}/reports/export?${searchParams.toString()}`,
      { responseType: 'blob' }
    )
    
    // Create download
    const blob = new Blob([response.data], { 
      type: response.headers['content-type'] || 'application/octet-stream' 
    })
    
    const url = window.URL.createObjectURL(blob)
    const link = document.createElement('a')
    link.href = url
    link.download = `accounting-report-${params.type}-${new Date().toISOString().split('T')[0]}.${params.format}`
    document.body.appendChild(link)
    link.click()
    document.body.removeChild(link)
    window.URL.revokeObjectURL(url)

    return blob
  }

  /**
   * Export all reports
   */
  async exportAllReports(params: { format?: string; filters?: any }): Promise<Blob> {
    const searchParams = new URLSearchParams()
    
    if (params.format) searchParams.append('format', params.format)
    if (params.filters) {
      Object.entries(params.filters).forEach(([key, value]) => {
        if (value !== null && value !== undefined && value !== '') {
          searchParams.append(key, value.toString())
        }
      })
    }

    const response = await apiClient.get(
      `${this.baseUrl}/reports/export-all?${searchParams.toString()}`,
      { responseType: 'blob' }
    )
    
    // Create download
    const blob = new Blob([response.data], { 
      type: response.headers['content-type'] || 'application/octet-stream' 
    })
    
    const url = window.URL.createObjectURL(blob)
    const link = document.createElement('a')
    link.href = url
    link.download = `accounting-all-reports-${new Date().toISOString().split('T')[0]}.${params.format || 'xlsx'}`
    document.body.appendChild(link)
    link.click()
    document.body.removeChild(link)
    window.URL.revokeObjectURL(url)

    return blob
  }

  // =====================================================================
  // Health Check
  // =====================================================================

  /**
   * Check accounting module health
   */
  async healthCheck(): Promise<{ status: string; module: string; timestamp: string }> {
    const response = await apiClient.get(`${this.baseUrl}/health`)
    return response.data
  }

  // =====================================================================
  // Utility Methods
  // =====================================================================

  /**
   * Build query parameters for API calls
   */
  private buildQueryParams(params: Record<string, any>): URLSearchParams {
    const searchParams = new URLSearchParams()
    
    Object.entries(params).forEach(([key, value]) => {
      if (value !== null && value !== undefined && value !== '') {
        if (Array.isArray(value)) {
          value.forEach(item => searchParams.append(key, item.toString()))
        } else {
          searchParams.append(key, value.toString())
        }
      }
    })
    
    return searchParams
  }

  /**
   * Format date for API calls
   */
  private formatDate(date: Date | string): string {
    if (typeof date === 'string') return date
    return date.toISOString().split('T')[0]
  }

  /**
   * Format datetime for API calls
   */
  private formatDateTime(date: Date | string): string {
    if (typeof date === 'string') return date
    return date.toISOString()
  }

  /**
   * Parse API error response
   */
  private parseErrorResponse(error: any): string {
    if (error.response?.data?.detail) {
      return Array.isArray(error.response.data.detail) 
        ? error.response.data.detail[0].msg 
        : error.response.data.detail
    }
    
    if (error.response?.data?.message) {
      return error.response.data.message
    }
    
    if (error.message) {
      return error.message
    }
    
    return 'An unknown error occurred'
  }
}

// Create and export singleton instance
export const accountingService = new AccountingService()

// Export service class for testing
export { AccountingService as AccountingServiceClass }

// Export default
export default accountingService